packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -68,55 +70,101 @@ library LibProposing {
         bytes calldata _txList
     )
         internal
-        returns (TaikoData.BlockMetadata memory meta_, TaikoData.EthDeposit[] memory deposits_)
+        returns (TaikoData.BlockMetadata2 memory meta_, TaikoData.EthDeposit[] memory deposits_)
     {
-        Local memory local;
-        local.params = abi.decode(_data, (TaikoData.BlockParams));
-
-        if (local.params.coinbase == address(0)) {
-            local.params.coinbase = msg.sender;
-        }
-
         // Taiko, as a Based Rollup, enables permissionless block proposals.
+        Local memory local;
         local.b = _state.slotB;
+        local.postFork = local.b.numBlocks >= _config.forkHeight;
 
         // It's essential to ensure that the ring buffer for proposed blocks
         // still has space for at least one more block.
         if (local.b.numBlocks >= local.b.lastVerifiedBlockId + _config.blockMaxProposals + 1) {
             revert L1_TOO_MANY_BLOCKS();
         }
 
-        local.parentMetaHash =
-            _state.blocks[(local.b.numBlocks - 1) % _config.blockRingBufferSize].metaHash;
-        // assert(parentMetaHash != 0);
+        TaikoData.BlockParams2 memory params;
+
+        if (local.postFork) {
+            if (_data.length != 0) {
+                params = abi.decode(_data, (TaikoData.BlockParams2));
+                // otherwise use a default BlockParams2 with 0 values
+            }
+        } else {
+            params = LibData.paramV1toV2(abi.decode(_data, (TaikoData.BlockParams)));
+        }
+
+        if (params.coinbase == address(0)) {
+            params.coinbase = msg.sender;
+        }
+
+        if (!local.postFork || params.anchorBlockId == 0) {
+            params.anchorBlockId = uint64(block.number - 1);
+        }
+
+        if (!local.postFork || params.timestamp == 0) {
+            params.timestamp = uint64(block.timestamp);
+        }
 
-        // Check if parent block has the right meta hash. This is to allow the proposer to make sure
-        // the block builds on the expected latest chain state.
-        if (local.params.parentMetaHash != 0 && local.parentMetaHash != local.params.parentMetaHash)
+        // Verify params against the parent block.
         {
-            revert L1_UNEXPECTED_PARENT();
+            TaikoData.Block storage parentBlk =
+                _state.blocks[(local.b.numBlocks - 1) % _config.blockRingBufferSize];
+
+            // Verify the passed in L1 state block number.
+            // We only allow the L1 block to be 2 epochs old.
+            // The other constraint is that the L1 block number needs to be larger than or equal the
+            // one
+            // in the previous L2 block.
+            if (
+                params.anchorBlockId + 64 < block.number //
+                    || params.anchorBlockId >= block.number
+                    || params.anchorBlockId < parentBlk.anchorBlockId
+            ) {
+                revert L1_INVALID_ANCHOR_BLOCK();
```

Comment:
Out of curiosity. This 2-epoch check is to have a "foreseeable" timeframe in which we can confidently say the preconfers will include L2 blocks for sure.
But what happens in case on L2 we execute, but (for whatever reason) we skip 64 L1 blocks in the meantime (system downtime, wtwr) and no proposal happens on L1 ? In such case those L2 blocks cannot be proposed anymore.. (?) all of those blocks are not valid and TaikoL2 needs to have a reorg, right ? Proposer needs to propose with `anchorBlockId == 0` to be able to pass this check but the `not-yet-proposed-but-preconfed-txns` (within this 64 slots, which was missed) are gone.
Should not we need a failsafe mechanism or that is the nature of this and we are OK with that ?

Reply:
> In such case those L2 blocks cannot be proposed anymore.. 

Even if there is a long time, much longer than 64 blocks, on L1 without any L2 blocks proposed, proposers can still propose new blocks.

Reply:
Yep, somebody just needs to build a block with more recent L1 data.

Reply:
Sure i meant, that the "promised" (preconfirmed TXN) L2 blocks are gone, but chain is operational!